src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
+                    if (res != 1)
+                    {
+                        throw new CryptographicException();
+                    }
+                }
+            }
+
+            internal void Evaluate(
+                DateTime verificationTime,
+                OidCollection applicationPolicy,
+                OidCollection certificatePolicy,
+                X509RevocationMode revocationMode,
+                X509RevocationFlag revocationFlag)
+            {
+                Debug.Assert(_chainContext != null);
+
+                long timeInMsFromUnixEpoch = new DateTimeOffset(verificationTime).ToUnixTimeMilliseconds();
+                _isValid = Interop.AndroidCrypto.X509ChainBuild(_chainContext, timeInMsFromUnixEpoch);
+                if (!_isValid)
+                {
+                    // Android always validates name, time, signature, and trusted root.
+                    // There is no way bypass that validation and build a path.
+                    ChainElements = Array.Empty<X509ChainElement>();
+
+                    Interop.AndroidCrypto.ValidationError[] errors = Interop.AndroidCrypto.X509ChainGetErrors(_chainContext);
+                    var chainStatus = new X509ChainStatus[errors.Length];
+                    for (int i = 0; i < errors.Length; i++)
+                    {
+                        Interop.AndroidCrypto.ValidationError error = errors[i];
+                        chainStatus[i] = ValidationErrorToChainStatus(error);
+                        Marshal.FreeHGlobal(error.Message);
+                    }
+
+                    ChainStatus = chainStatus;
+                    return;
+                }
+
+                if (revocationMode != X509RevocationMode.NoCheck)
+                {
+                    if (revocationFlag == X509RevocationFlag.EntireChain)
+                    {
+                        throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationFlag.{revocationFlag})");
+                    }
+
+                    if (!Interop.AndroidCrypto.X509ChainSupportsRevocationOptions())
+                    {
+                        if (revocationFlag == X509RevocationFlag.EndCertificateOnly)
+                        {
+                            // No way to specfiy end certificate only if revocation options are not available
+                            throw new PlatformNotSupportedException(SR.Format(SR.Chain_SettingNotSupported, $"{nameof(X509RevocationFlag)}.{nameof(X509RevocationFlag.EndCertificateOnly)}"));
+                        }
+
+                        // Defaults to offline when revocation options are not available
+                        if (revocationMode == X509RevocationMode.Online)
+                        {
+                            throw new NotImplementedException($"{nameof(Evaluate)} (X509RevocationMode.{revocationMode})");
+                        }
```

Comment:
If the options aren't available then there's too much revocation checking?  Not enough?  The logic here suggests both things.

If the default is no revocation checking, then I think the right answer is to just report RevocationStatusUnknown.

If the default is something else I might have other opinions.

Reply:
If the options aren't available, we can still turn revocation checking on or off. By default, it should be offline and exclude the root when turned on. From docs I've seen, we should still be able to enable/disable OCSP though setting the `ocsp.enable` security property: https://docs.oracle.com/javase/8/docs/technotes/guides/security/certpath/CertPathProgGuide.html#AppC
So I think the only thing that would truly be unsupported without the options is `X509RevocationFlag.EndCertificateOnly` (and `EntireChain`).

This PR only turns revocation checking on/off and doesn't handle online mode without the options API available (largely because I haven't set up my local environment with an version that doesn't have the options yet). My intent was to make a separate/follow-up change to remove this NotImplementedException and handle online mode in this case (assuming the property works the same on Android).

Reply:
I'm having trouble wading through the factory and such handwaviness in Java/Android's documentation here, but the notion that CRL/OCSP fetch matches our Offline mode seems weird.  (There's a cache, if it's in the cache, great, if not, you'll get RevocationStatusUnknown).

Our Online mode is "use the cache, if you can, but you're allowed to go download things if you have to".

Reply:
> My intent was to make a separate/follow-up change to remove this NotImplementedException and handle online mode in this case (assuming the property works the same on Android).

If that's the path forward, please create an issue to track the work, so that if you get pulled onto a different project we remember it's still pending :smile:

Reply:
Rewrote https://github.com/dotnet/runtime/issues/45741 to be more reflective of the current direction (instead of the just P/Invokes one). I'm going with using that as the master work-tracking issue instead of multiple issues, since it is all this work is still on-going.